### PR TITLE
Add support for better timing with DroneCAN GPS

### PR DESCRIPTION
**Pull request by bugobliterator**

### CODE REVIEW

File: Tools/AP_Periph/AP_Periph.h [hunk 1]
```diff
@@ -247,6 +247,7 @@ class AP_Periph_FW {
     uint32_t last_gps_update_ms;
     uint32_t last_baro_update_ms;
     uint32_t last_airspeed_update_ms;
+    bool saw_lock_once;
```
Comment (nkruzan) [inline Tools/AP_Periph/AP_Periph.h:250]: nitpick - I'd rather see this as saw_gps_lock_once

File: libraries/AP_GPS/AP_GPS.cpp [hunk 2]
```diff
@@ -861,11 +879,12 @@ void AP_GPS::update_instance(uint8_t instance)
...
     } else {
-        if (state[instance].uart_timestamp_ms != 0) {
+        if (state[instance].corrected_timestamp_updated) {
             // set the timestamp for this messages based on
-            // set_uart_timestamp() in backend, if available
-            tnow = state[instance].uart_timestamp_ms;
-            state[instance].uart_timestamp_ms = 0;
+            // set_uart_timestamp() or per specific transport in backend
+            // , if available
+            tnow = state[instance].last_corrected_gps_time_us/1000;
```
Comment (tridge) [inline libraries/AP_GPS/AP_GPS.cpp:886]: 1000U

File: libraries/AP_GPS/AP_GPS_UAVCAN.cpp [hunk 3]
```diff
@@ -578,8 +583,41 @@ void AP_GPS_UAVCAN::handle_fix2_msg(const Fix2Cb &cb)
...
+        interim_state.last_corrected_gps_time_us = jitter_correction.correct_offboard_timestamp_usec(cb.msg->timestamp.usec, (cb.msg->getUtcTimestamp().toUSec() + NATIVE_TIME_OFFSET));
+        interim_state.last_gps_time_ms = interim_state.last_corrected_gps_time_us/1000;
+        interim_state.last_corrected_gps_time_us -= cb.msg->timestamp.usec - cb.msg->gnss_timestamp.usec;
+        // this is also the time the message was received on the UART on other end.
+        interim_state.corrected_timestamp_updated = true;
+    } else {
+        interim_state.last_gps_time_ms = jitter_correction.correct_offboard_timestamp_usec(cb.msg->timestamp.usec, cb.msg->getUtcTimestamp().toUSec() + NATIVE_TIME_OFFSET)/1000;
+    }
+
+#if 0
```
Comment (tridge) [inline libraries/AP_GPS/AP_GPS_UAVCAN.cpp:598]: maybe #if GPS_PPS_EMULATION

**Review by tridge — CHANGES_REQUESTED** [review]

File: libraries/AP_GPS/AP_GPS_UAVCAN.cpp [hunk 4]
```diff
@@ -578,8 +585,41 @@ void AP_GPS_UAVCAN::handle_fix2_msg(const Fix2Cb &cb)
...
+        //we are called from ISR context
+        chSysLockFromISR();
+        hal.gpio->toggle(51);
+        chSysUnlockFromISR();
+    };
+
+    static uint64_t next_toggle, last_toggle;
     
-    interim_state.last_gps_time_ms = AP_HAL::millis();
+    next_toggle = (cb.msg->timestamp.usec) + (1000000 - ((cb.msg->timestamp.usec) % 1000000));
```
Comment (tridge) [inline libraries/AP_GPS/AP_GPS_UAVCAN.cpp:615]: unsigned

File: libraries/AP_GPS/GPS_Backend.cpp [hunk 5]
```diff
@@ -287,21 +294,26 @@ void AP_GPS_Backend::check_new_itow(uint32_t itow, uint32_t msg_length)
...
+        state.last_corrected_gps_time_us = local_us;
+        state.corrected_timestamp_updated = true;
 
         // look for lagged data from the GPS. This is meant to detect
         // the case that the GPS is trying to push more data into the
         // UART than can fit (eg. with GPS_RAW_DATA at 115200).
         float expected_lag;
         if (gps.get_lag(state.instance, expected_lag)) {
-            float lag_s = (now - state.uart_timestamp_ms) * 0.001;
+            float lag_s = (now - (state.last_corrected_gps_time_us/1000)) * 0.001;
```
Comment (tridge) [inline libraries/AP_GPS/GPS_Backend.cpp:305]: unsigned

**Review by tridge — CHANGES_REQUESTED** [review]

**Comment by tridge** [comment]
happy to merge when the trivial changes are fixed and passes CI

**Review by tridge — APPROVED** [review]

**Pull request merged**